docs/arch/02-core-concepts.md

```diff
@@ -0,0 +1,702 @@
+
+**Related concepts:** Transport, Middleware, Session
+
+### Middleware
+
+**Middleware** is a composable layer in the request processing chain. Each middleware can inspect, modify, or reject requests.
+
+**Eight middleware types:**
+
+1. **Authentication** (`auth`) - JWT token validation
+2. **Token Exchange** (`tokenexchange`) - OAuth token exchange
+3. **MCP Parser** (`mcp-parser`) - JSON-RPC parsing
+4. **Tool Filter** (`tool-filter`) - Filter and override tools in `tools/list` responses
+5. **Tool Call Filter** (`tool-call-filter`) - Validate and map `tools/call` requests
+6. **Telemetry** (`telemetry`) - OpenTelemetry instrumentation
+7. **Authorization** (`authorization`) - Cedar policy evaluation
+8. **Audit** (`audit`) - Request logging
+
+**Execution order (request flow):**
+Middleware applied in reverse configuration order. Requests flow through: Audit* → Authorization* → Telemetry* → Parser → Token Exchange* → Auth → Tool Call Filter* → Tool Filter* → MCP Server
+
+(*optional middleware, only present if configured)
+
+**Implementation:**
+- Interface: `pkg/transport/types/transport.go`
+- Factory: `pkg/runner/middleware.go`
+- Documentation: `docs/middleware.md`
+
+**Related concepts:** Proxy, Authentication, Authorization
+
+### RunConfig
+
+**RunConfig** is ToolHive's standard configuration format for running MCP servers. It's a JSON/YAML structure that contains everything needed to deploy a workload.
+
+**Configuration categories:**
+- **Execution**: `image`, `cmdArgs`, `transport`, `name`, `containerName`
+- **Networking**: `host`, `port`, `targetPort`, `targetHost`, `isolateNetwork`, `proxyMode`
+- **Security**: `permissionProfile`, `secrets`, `oidcConfig`, `authzConfig`, `trustProxyHeaders`
+- **Observability**: `auditConfig`, `telemetryConfig`, `debug`
+- **Customization**: `envVars`, `volumes`, `toolsFilter`, `toolsOverride`, `ignoreConfig`
+- **Organization**: `group`, `containerLabels`
+- **Middleware**: `middlewareConfigs` - Dynamic middleware chain configuration
+- **Remote servers**: `remoteURL`, `remoteAuthConfig`
+- **Kubernetes**: `k8sPodTemplatePatch`
+
+See `pkg/runner/config.go` for complete field reference.
+
+**Schema version:** `v0.1.0` (current)
+
+**Portability:**
+- Export: `thv export <workload>` → JSON file
+- Import: `thv run --from-config <file>`
+- API contract: Format is versioned and stable
+
+**Implementation:**
+- Definition: `pkg/runner/config.go`
+- Schema version: `pkg/runner/config.go`
+
+**Related concepts:** Workload, Permission Profile, Middleware
+
+### Permission Profile
+
+A **permission profile** defines security boundaries for MCP servers:
+
+**Three permission types:**
+
+1. **File System Access**:
+   - `read` - Mount paths as read-only
+   - `write` - Mount paths as read-write
+   - Mount declaration formats: `path`, `host:container`, `uri://resource:container`
+
+2. **Network Access**:
+   - `outbound.insecureAllowAll` - Allow all outbound connections
```

Comment:
Network access field names are incorrect. The code uses **snake_case**, not camelCase.

```suggestion
   - `outbound.insecure_allow_all` - Allow all outbound connections
   - `outbound.allow_host` - Whitelist specific hosts
   - `outbound.allow_port` - Whitelist specific ports
   - `inbound.allow_host` - Whitelist inbound connections
```

Evidence: pkg/permissions/profile.go:59-72 (OutboundNetworkPermissions and InboundNetworkPermissions structs use snake_case in JSON/YAML tags)